lxd/instance/drivers/driver_qemu_templates.go

```diff
@@ -60,68 +60,72 @@ size = "{{.memSizeBytes}}B"
 
 var qemuSerial = template.Must(template.New("qemuSerial").Parse(`
 # LXD serial identifier
-[device]
-driver = "virtio-serial"
-
-[device]
-driver = "virtserialport"
-name = "org.linuxcontainers.lxd"
-chardev = "vserial"
+[device "dev-qemu_serial"]
+{{- if eq .bus "pci" "pcie"}}
+driver = "virtio-serial-pci"
+bus = "{{.devBus}}"
+addr = "{{.devAddr}}"
+{{end -}}
+{{if eq .bus "ccw" -}}
+driver = "virtio-scsi-ccw"
+{{end -}}
+{{if .multifunction -}}
+multifunction = "on"
+{{- end }}
 
-[chardev "vserial"]
+[chardev "qemu_serial-chardev"]
 backend = "ringbuf"
 size = "{{.ringbufSizeBytes}}B"
+
+[device "qemu_serial"]
+driver = "virtserialport"
+name = "org.linuxcontainers.lxd"
+chardev = "qemu_serial-chardev"
+bus = "dev-qemu_serial.0"
 `))
 
-var qemuSCSI = template.Must(template.New("qemuSCSI").Parse(`
-# SCSI controller
-{{- if eq .architecture "x86_64" "aarch64" }}
-[device "qemu_pcie1"]
+var qemuPCIe = template.Must(template.New("qemuPCIe").Parse(`
+[device "qemu_pcie{{.index}}"]
 driver = "pcie-root-port"
-port = "0x10"
-chassis = "1"
 bus = "pcie.0"
+addr = "{{.addr}}"
+port = "{{.index}}"
+chassis = "{{.index}}"
+{{if .multifunction -}}
 multifunction = "on"
-addr = "0x2"
 {{- end }}
+`))
 
+var qemuSCSI = template.Must(template.New("qemuSCSI").Parse(`
+# SCSI controller
 [device "qemu_scsi"]
-{{- if ne .architecture "s390x"}}
+{{- if eq .bus "pci" "pcie"}}
 driver = "virtio-scsi-pci"
-{{- if eq .architecture "ppc64le" }}
-bus = "pci.0"
-{{- else}}
-bus = "qemu_pcie1"
-addr = "0x0"
-{{- end}}
-{{- else}}
+bus = "{{.devBus}}"
+addr = "{{.devAddr}}"
+{{end -}}
+{{if eq .bus "ccw" -}}
 driver = "virtio-scsi-ccw"
-{{- end}}
+{{end -}}
+{{if .multifunction -}}
+multifunction = "on"
+{{- end }}
 `))
 
 var qemuBalloon = template.Must(template.New("qemuBalloon").Parse(`
 # Balloon driver
-{{- if eq .architecture "x86_64" "aarch64" }}
-[device "qemu_pcie2"]
-driver = "pcie-root-port"
-port = "0x11"
-chassis = "2"
-bus = "pcie.0"
-addr = "0x2.0x1"
-{{- end }}
-
 [device "qemu_ballon"]
```

Comment:
@stgraber just spotted this typo in the existing device name.

Reply:
Fixed